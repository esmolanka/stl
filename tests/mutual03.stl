type Even = < EZero | ESucc : Odd | ESuccSucc : Even >
  with Odd = < OZero | OSucc : Even | OSuccSucc : Odd >

#eval Even

#eval Odd

#check <EZero | ESucc : Odd | ESuccSucc : Even> <: Even
#check Even <: <EZero | ESucc : Odd | ESuccSucc : Even>

#check <OZero | OSucc : Even | OSuccSucc : Odd> <: Odd
#check Odd <: <OZero | OSucc : Even | OSuccSucc : Odd>

