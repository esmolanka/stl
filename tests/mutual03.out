tests/mutual03.types:4:1-4:11:
  Normalised:
    μ Even. ∀ (ρ : Row).
      < EZero? : Unit
      | ESucc? :
          μ Odd. ∀ (ρ : Row).
            <OZero? : Unit | OSucc? : Even | OSuccSucc? : Odd | ... ρ>
      | ESuccSucc? : Even
      | ... ρ
      >
  Kind:
    Type
tests/mutual03.types:6:1-6:10:
  Normalised:
    μ Odd. ∀ (ρ : Row).
      < OZero? : Unit
      | OSucc? :
          μ Even. ∀ (ρ : Row).
            <EZero? : Unit | ESucc? : Odd | ESuccSucc? : Even | ... ρ>
      | OSuccSucc? : Odd
      | ... ρ
      >
  Kind:
    Type
tests/mutual03.types:8:1-8:56:
  Subtype:
    ∀ (ρ : Row).
      < EZero? : Unit
      | ESucc? :
          μ Odd. ∀ (ρ : Row).
            < OZero? : Unit
            | OSucc? :
                μ Even. ∀ (ρ : Row).
                  <EZero? : Unit | ESucc? : Odd | ESuccSucc? : Even | ... ρ>
            | OSuccSucc? : Odd
            | ... ρ
            >
      | ESuccSucc? :
          μ Even. ∀ (ρ : Row).
            < EZero? : Unit
            | ESucc? :
                μ Odd. ∀ (ρ : Row).
                  <OZero? : Unit | OSucc? : Even | OSuccSucc? : Odd | ... ρ>
            | ESuccSucc? : Even
            | ... ρ
            >
      | ... ρ
      >
  Supertype:
    μ Even. ∀ (ρ : Row).
      < EZero? : Unit
      | ESucc? :
          μ Odd. ∀ (ρ : Row).
            <OZero? : Unit | OSucc? : Even | OSuccSucc? : Odd | ... ρ>
      | ESuccSucc? : Even
      | ... ρ
      >
  Result:
    OK
tests/mutual03.types:9:1-9:56:
  Subtype:
    μ Even. ∀ (ρ : Row).
      < EZero? : Unit
      | ESucc? :
          μ Odd. ∀ (ρ : Row).
            <OZero? : Unit | OSucc? : Even | OSuccSucc? : Odd | ... ρ>
      | ESuccSucc? : Even
      | ... ρ
      >
  Supertype:
    ∀ (ρ : Row).
      < EZero? : Unit
      | ESucc? :
          μ Odd. ∀ (ρ : Row).
            < OZero? : Unit
            | OSucc? :
                μ Even. ∀ (ρ : Row).
                  <EZero? : Unit | ESucc? : Odd | ESuccSucc? : Even | ... ρ>
            | OSuccSucc? : Odd
            | ... ρ
            >
      | ESuccSucc? :
          μ Even. ∀ (ρ : Row).
            < EZero? : Unit
            | ESucc? :
                μ Odd. ∀ (ρ : Row).
                  <OZero? : Unit | OSucc? : Even | OSuccSucc? : Odd | ... ρ>
            | ESuccSucc? : Even
            | ... ρ
            >
      | ... ρ
      >
  Result:
    OK
tests/mutual03.types:11:1-11:55:
  Subtype:
    ∀ (ρ : Row).
      < OZero? : Unit
      | OSucc? :
          μ Even. ∀ (ρ : Row).
            < EZero? : Unit
            | ESucc? :
                μ Odd. ∀ (ρ : Row).
                  <OZero? : Unit | OSucc? : Even | OSuccSucc? : Odd | ... ρ>
            | ESuccSucc? : Even
            | ... ρ
            >
      | OSuccSucc? :
          μ Odd. ∀ (ρ : Row).
            < OZero? : Unit
            | OSucc? :
                μ Even. ∀ (ρ : Row).
                  <EZero? : Unit | ESucc? : Odd | ESuccSucc? : Even | ... ρ>
            | OSuccSucc? : Odd
            | ... ρ
            >
      | ... ρ
      >
  Supertype:
    μ Odd. ∀ (ρ : Row).
      < OZero? : Unit
      | OSucc? :
          μ Even. ∀ (ρ : Row).
            <EZero? : Unit | ESucc? : Odd | ESuccSucc? : Even | ... ρ>
      | OSuccSucc? : Odd
      | ... ρ
      >
  Result:
    OK
tests/mutual03.types:12:1-12:55:
  Subtype:
    μ Odd. ∀ (ρ : Row).
      < OZero? : Unit
      | OSucc? :
          μ Even. ∀ (ρ : Row).
            <EZero? : Unit | ESucc? : Odd | ESuccSucc? : Even | ... ρ>
      | OSuccSucc? : Odd
      | ... ρ
      >
  Supertype:
    ∀ (ρ : Row).
      < OZero? : Unit
      | OSucc? :
          μ Even. ∀ (ρ : Row).
            < EZero? : Unit
            | ESucc? :
                μ Odd. ∀ (ρ : Row).
                  <OZero? : Unit | OSucc? : Even | OSuccSucc? : Odd | ... ρ>
            | ESuccSucc? : Even
            | ... ρ
            >
      | OSuccSucc? :
          μ Odd. ∀ (ρ : Row).
            < OZero? : Unit
            | OSucc? :
                μ Even. ∀ (ρ : Row).
                  <EZero? : Unit | ESucc? : Odd | ESuccSucc? : Even | ... ρ>
            | OSuccSucc? : Odd
            | ... ρ
            >
      | ... ρ
      >
  Result:
    OK